### ISSUE TITLE
Got AttributeError: module 'scipy' has no attribute 'special' when building scipy with `special` import

### ISSUE THREAD

**Issue by charlotte12l** [serge-sans-paille/pythran#1815]
I'm trying to use Pythran to improve a scipy function `_count_paths_outside_method`. My SciPy branch is https://github.com/charlotte12l/scipy/tree/pythran_count.  If I use Pythran to compile the single file `scipy/stats/_count_paths_outside_method_pythran.py`, it would work fine:

```
(scipydev) charlotte@CHARLOTLIU-MB0 scipy % pythran scipy/stats/_count_paths_outside_method_pythran.py 
WARNING: Pythran support for package 'scipy' will be reduced: this module is not available at runtime.
WARNING: Pythran support for package 'scipy.special' will be reduced: this module is not available at runtime.
```

However, if I use SciPy's building system to compile the file, there would be many errors. 

If I installed Pythran master branch and didn't uninstalled scipy , it would build successfully but would throw a `bus error` when testing `pytest scipy/stats/tests/test_stats.py::TestKSTwoSamples::test100_110`. If I installed Pythran `feature/break-scipy-loop` branch and pip uninstalled scipy, it can't build successfully:

```
building extension "scipy.stats._sobol" sources
building extension "scipy.stats._qmc_cy" sources
building extension "scipy.stats.biasedurn" sources
Traceback (most recent call last):
  File "setup.py", line 631, in <module>
    setup_package()
  File "setup.py", line 627, in setup_package
    setup(**metadata)
  File "/Users/charlotte/.local/lib/python3.8/site-packages/numpy-1.21.0rc2-py3.8-macosx-10.9-x86_64.egg/numpy/distutils/core.py", line 169, in setup
    return old_setup(**new_attr)
  File "/Users/charlotte/anaconda3/envs/scipydev/lib/python3.8/site-packages/setuptools/__init__.py", line 153, in setup
    return distutils.core.setup(**attrs)
  File "/Users/charlotte/anaconda3/envs/scipydev/lib/python3.8/distutils/core.py", line 148, in setup
    dist.run_commands()
  File "/Users/charlotte/anaconda3/envs/scipydev/lib/python3.8/distutils/dist.py", line 966, in run_commands
    self.run_command(cmd)
  File "/Users/charlotte/anaconda3/envs/scipydev/lib/python3.8/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "/Users/charlotte/.local/lib/python3.8/site-packages/numpy-1.21.0rc2-py3.8-macosx-10.9-x86_64.egg/numpy/distutils/command/build_ext.py", line 101, in run
    self.run_command('build_src')
  File "/Users/charlotte/anaconda3/envs/scipydev/lib/python3.8/distutils/cmd.py", line 313, in run_command
    self.distribution.run_command(command)
  File "/Users/charlotte/anaconda3/envs/scipydev/lib/python3.8/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "/Users/charlotte/.local/lib/python3.8/site-packages/numpy-1.21.0rc2-py3.8-macosx-10.9-x86_64.egg/numpy/distutils/command/build_src.py", line 144, in run
    self.build_sources()
  File "/Users/charlotte/.local/lib/python3.8/site-packages/numpy-1.21.0rc2-py3.8-macosx-10.9-x86_64.egg/numpy/distutils/command/build_src.py", line 161, in build_sources
    self.build_extension_sources(ext)
  File "/Users/charlotte/.local/lib/python3.8/site-packages/numpy-1.21.0rc2-py3.8-macosx-10.9-x86_64.egg/numpy/distutils/command/build_src.py", line 306, in build_extension_sources
    sources = list(ext.sources)
  File "/Users/charlotte/.local/lib/python3.8/site-packages/pythran-0.9.12.dev0-py3.8.egg/pythran/dist.py", line 170, in sources
    tc.compile_pythranfile(source, output_file,
  File "/Users/charlotte/.local/lib/python3.8/site-packages/pythran-0.9.12.dev0-py3.8.egg/pythran/toolchain.py", line 472, in compile_pythranfile
    output_file = compile_pythrancode(module_name, open(file_path).read(),
  File "/Users/charlotte/.local/lib/python3.8/site-packages/pythran-0.9.12.dev0-py3.8.egg/pythran/toolchain.py", line 394, in compile_pythrancode
    module, error_checker = generate_cxx(module_name, pythrancode, specs, opts,
  File "/Users/charlotte/.local/lib/python3.8/site-packages/pythran-0.9.12.dev0-py3.8.egg/pythran/toolchain.py", line 138, in generate_cxx
    pm, ir, docstrings = front_middle_end(module_name, code, optimizations,
  File "/Users/charlotte/.local/lib/python3.8/site-packages/pythran-0.9.12.dev0-py3.8.egg/pythran/toolchain.py", line 105, in front_middle_end
    refine(pm, ir, optimizations)
  File "/Users/charlotte/.local/lib/python3.8/site-packages/pythran-0.9.12.dev0-py3.8.egg/pythran/middlend.py", line 61, in refine
    apply_optimisation |= pm.apply(optimization, node)[0]
  File "/Users/charlotte/.local/lib/python3.8/site-packages/pythran-0.9.12.dev0-py3.8.egg/pythran/passmanager.py", line 236, in apply
    return a.apply(node)
  File "/Users/charlotte/.local/lib/python3.8/site-packages/pythran-0.9.12.dev0-py3.8.egg/pythran/passmanager.py", line 199, in apply
    new_node = self.run(node)
  File "/Users/charlotte/.local/lib/python3.8/site-packages/pythran-0.9.12.dev0-py3.8.egg/pythran/passmanager.py", line 187, in run
    n = super(Transformation, self).run(node)
  File "/Users/charlotte/.local/lib/python3.8/site-packages/pythran-0.9.12.dev0-py3.8.egg/pythran/passmanager.py", line 96, in run
    return self.visit(node)
  File "/Users/charlotte/.local/lib/python3.8/site-packages/pythran-0.9.12.dev0-py3.8.egg/pythran/passmanager.py", line 78, in visit
    return super(ContextManager, self).visit(node)
  File "/Users/charlotte/anaconda3/envs/scipydev/lib/python3.8/ast.py", line 371, in visit
    return visitor(node)
  File "/Users/charlotte/.local/lib/python3.8/site-packages/pythran-0.9.12.dev0-py3.8.egg/pythran/optimizations/constant_folding.py", line 116, in generic_visit
    return Transformation.generic_visit(self, node)
  File "/Users/charlotte/anaconda3/envs/scipydev/lib/python3.8/ast.py", line 447, in generic_visit
    value = self.visit(value)
  File "/Users/charlotte/.local/lib/python3.8/site-packages/pythran-0.9.12.dev0-py3.8.egg/pythran/passmanager.py", line 78, in visit
    return super(ContextManager, self).visit(node)
  File "/Users/charlotte/anaconda3/envs/scipydev/lib/python3.8/ast.py", line 371, in visit
    return visitor(node)
  File "/Users/charlotte/.local/lib/python3.8/site-packages/pythran-0.9.12.dev0-py3.8.egg/pythran/optimizations/constant_folding.py", line 116, in generic_visit
    return Transformation.generic_visit(self, node)
  File "/Users/charlotte/anaconda3/envs/scipydev/lib/python3.8/ast.py", line 447, in generic_visit
    value = self.visit(value)
  File "/Users/charlotte/.local/lib/python3.8/site-packages/pythran-0.9.12.dev0-py3.8.egg/pythran/passmanager.py", line 78, in visit
    return super(ContextManager, self).visit(node)
  File "/Users/charlotte/anaconda3/envs/scipydev/lib/python3.8/ast.py", line 371, in visit
    return visitor(node)
  File "/Users/charlotte/.local/lib/python3.8/site-packages/pythran-0.9.12.dev0-py3.8.egg/pythran/optimizations/constant_folding.py", line 116, in generic_visit
    return Transformation.generic_visit(self, node)
  File "/Users/charlotte/anaconda3/envs/scipydev/lib/python3.8/ast.py", line 447, in generic_visit
    value = self.visit(value)
  File "/Users/charlotte/.local/lib/python3.8/site-packages/pythran-0.9.12.dev0-py3.8.egg/pythran/passmanager.py", line 78, in visit
    return super(ContextManager, self).visit(node)
  File "/Users/charlotte/anaconda3/envs/scipydev/lib/python3.8/ast.py", line 371, in visit
    return visitor(node)
  File "/Users/charlotte/.local/lib/python3.8/site-packages/pythran-0.9.12.dev0-py3.8.egg/pythran/optimizations/constant_folding.py", line 116, in generic_visit
    return Transformation.generic_visit(self, node)
  File "/Users/charlotte/anaconda3/envs/scipydev/lib/python3.8/ast.py", line 456, in generic_visit
    new_node = self.visit(old_value)
  File "/Users/charlotte/.local/lib/python3.8/site-packages/pythran-0.9.12.dev0-py3.8.egg/pythran/passmanager.py", line 78, in visit
    return super(ContextManager, self).visit(node)
  File "/Users/charlotte/anaconda3/envs/scipydev/lib/python3.8/ast.py", line 371, in visit
    return visitor(node)
  File "/Users/charlotte/.local/lib/python3.8/site-packages/pythran-0.9.12.dev0-py3.8.egg/pythran/optimizations/constant_folding.py", line 116, in generic_visit
    return Transformation.generic_visit(self, node)
  File "/Users/charlotte/anaconda3/envs/scipydev/lib/python3.8/ast.py", line 447, in generic_visit
    value = self.visit(value)
  File "/Users/charlotte/.local/lib/python3.8/site-packages/pythran-0.9.12.dev0-py3.8.egg/pythran/passmanager.py", line 78, in visit
    return super(ContextManager, self).visit(node)
  File "/Users/charlotte/anaconda3/envs/scipydev/lib/python3.8/ast.py", line 371, in visit
    return visitor(node)
  File "/Users/charlotte/.local/lib/python3.8/site-packages/pythran-0.9.12.dev0-py3.8.egg/pythran/optimizations/constant_folding.py", line 116, in generic_visit
    return Transformation.generic_visit(self, node)
  File "/Users/charlotte/anaconda3/envs/scipydev/lib/python3.8/ast.py", line 456, in generic_visit
    new_node = self.visit(old_value)
  File "/Users/charlotte/.local/lib/python3.8/site-packages/pythran-0.9.12.dev0-py3.8.egg/pythran/passmanager.py", line 78, in visit
    return super(ContextManager, self).visit(node)
  File "/Users/charlotte/anaconda3/envs/scipydev/lib/python3.8/ast.py", line 371, in visit
    return visitor(node)
  File "/Users/charlotte/.local/lib/python3.8/site-packages/pythran-0.9.12.dev0-py3.8.egg/pythran/optimizations/constant_folding.py", line 84, in generic_visit
    value = eval(code, self.env)
  File "<constant folding>", line 20, in <module>
AttributeError: module 'scipy' has no attribute 'special'
```
 
I think it is related to scipy's building system& Pythran. I'm using MacOS Darwin 19.6.0.  Here is how to reproduce the error:

### 
I set up scipy env following https://github.com/scipy/scipy/blob/master/doc/source/dev/contributor/quickstart_mac.rst, here are the main steps:

1. Enter ``conda create --name scipydev python=3.8`` to create an empty virtual environment named ``scipydev`` 

2. Enter ``conda config --env --add channels conda-forge`` to tell Anaconda the source we want for our packages. Then enter ``conda install python=3.8 numpy pybind11 cython pythran pytest compilers sphinx pydata-sphinx-theme sphinx-panels matplotlib mypy`` to install the following packages:

   * ``numpy pybind11 cython pythran`` are four packages that SciPy depends on.

   * ``compilers`` holds compilers used to build SciPy's Fortran, C, and C++ source code.

   * ``pytest`` is needed for running the test suite.

   * ``sphinx``, ``pydata-sphinx-theme``, ``sphinx-panels`` and ``matplotlib`` are required to render the SciPy documentation.

   * ``mypy`` is a static type checker for Python. Consider using it.
3.  I once encountered `fatal error: 'cblas.h' file not found` so I also did ` conda install mkl ` and  `conda install blas`
4. Because I need Pythran `feature/break-scipy-loop` branch, so I did the following Pythran setups
   *  ``conda create --name pythrandev python=3.8``(actually I'm not sure whether it is necessary to create a new env )
   *   Go to a folder and do `git clone https://github.com/serge-sans-paille/pythran`
   * `cd pythran`
   *  `git checkout feature/break-scipy-loop`
   *   `python setup.py install --user`,  the packages would be installed  in `/Users/charlotte/.local/lib/python3.8/site-packages/`
   *   add `export PATH=$PATH:$HOME/.local/bin` in `~/.zshrc`
5.  `conda activate scipydev` to return to scipy environment. If you do `import pythran` you will find it imports the pythran from `/Users/charlotte/.local/lib/python3.8/site-packages/` rather than from the conda env. (But I don't think it will be a problem?)
6. `pip uninstall scipy` to remove the scipy that was installed when building pythran
4.  `git clone https://github.com/charlotte12l/scipy.git`
5.  In the terminal, navigate into the ``scipy`` root directory (e.g. ``cd scipy``). 
6. `git checkout pythran_count` to checkout my branch

7. Initialize git submodules: ``git submodule update --init``.

8. Do an in-place build: enter ``python3 setup.py build_ext --inplace``. And you will see the `AttributeError: module 'scipy' has no attribute 'special'`
9. If you build successfully, enter ``conda develop .`` . Then do `pytest scipy/stats/tests/test_stats.py::TestKSTwoSamples::test100_110`. to see if it will throw out a `bus error`.

Thank you very much!!!

cc @rgommers

**Comment by rgommers** [comment]
I think this should work
```
# In the scipy repo, edit `environment.yml` - remove `pythran` from it.Then:
conda env create -f environment.yml
conda activate scipy-dev
# Now switch to the Pythran repo and check out https://github.com/serge-sans-paille/pythran/pull/1810. Then:
pip install .   # Installs Pythran and its dependencies into the scipy-dev conda env
# Now switch back to the SciPy repo
python setup.py build_ext -i  #  This should work now.
```

**Comment by charlotte12l** [comment]
```conda env remove -n scipy-dev
rm -rf pythran
rm -rf scipy 
git clone https://github.com/charlotte12l/scipy.git
cd scipy
git checkout pythran_count
vim environment.yml 
conda env create -f environment.yml
conda activate scipy-dev
```

**Comment by rgommers** [comment]
Let me try and follow you exact steps above. They look good.

> Also, did you make sure you don't have other SciPy installed/ in the path?

Good question:) To check this, I usually do:
```
python -c "import scipy; print(scipy.__file__)"  # same for other packages
```
to be 100% sure about which package is actually imported.

**Comment by rgommers** [comment]
Problematic code block:
```python
            try:
                value = eval(code, self.env)
                new_node = to_ast(value)
                try:
                    if not ASTMatcher(node).search(new_node):
                        self.update = True
                        return new_node
                except DamnTooLongPattern as e:
                    print("W: ", e, " Assume no update happened.")
                return Transformation.generic_visit(self, node)
            except ConversionError as e:
                print('error in constant folding: ', e)
                raise
            except ToNotEval:
                return Transformation.generic_visit(self, node)
            except AttributeError as e:
                # FIXME union_ function is not handle by constant folding
                if "union" in e.args[0]:
                    return Transformation.generic_visit(self, node)
                elif "pythran" in e.args[0]:
                    # FIXME: Can be fix giving a Python implementation for
                    # these functions.
                    return Transformation.generic_visit(self, node)
                import IPython; IPython.embed()
                raise
```
(I added the `import IPython` line there). What it shows is that `AttributeError` is caught, and then re-raised:
```python
In [1]: e.args
Out[1]: ("module 'scipy' has no attribute 'special'",)
```
So an `elif` clause is missing there. Not sure exactly what it should be, maybe:
```python
elif "module" in e.args[0]:
    return Transformation.generic_visit(self, node)
```
?

**Comment by charlotte12l** [comment]
Thank you so much for your investigation!!!

> conda remove scipy   # NOTE: this step is critical!

> ```
> 
> The trouble with the `environment.yml` is `scikit-umfpack`. It's an optional dependency of SciPy, which in turn seems to pull in SciPy itself: https://github.com/conda-forge/scikit-umfpack-feedstock/blob/master/recipe/meta.yaml#L26
> 
> I will remove it from `environment.yml` to save us this headache.
> 

I'm not sure if it is a problem with  `scikit-umfpack`... because actually I didn't do `conda remove scipy` . I just pip uninstall scipy in `.local/lib/python3.8/`  before creating the environment.

**Comment by rgommers** [comment]
> Another interesting thing is that if you already have scipy in your path, then you can compile successfully and pass the test. However, last week, if I have scipy in `.local/lib/python3.8/`, then I can also compile successfully but failed the test with `bus error`. 

That makes sense. If you installed that `scipy` with `pip`, it will be compiled in a way that may not be compatible with the conda-installed Python and numpy. It's usually important to not mix pip and conda installed packages; if it's pure Python code it'll be fine, but when you have C/C++ extensions it's very fragile.

> So, to reproduce your successful test, I guess I need to already have a `scipy` in my conda env, and after the compilation, do `export PYTHONPATH=~/code/scipy`, and do `pytest scipy/stats/tests/test_stats.py::TestKSTwoSamples::test100_110`?

Yes exactly. That will work, because during the in-place scipy build, Pythran seems to be picking up the conda-installed scipy.

> I'm not sure if it is a problem with `scikit-umfpack`

It's anyway a good idea to get rid of it - already done in https://github.com/scipy/scipy/pull/14285.

**Comment by charlotte12l** [comment]
weird...  I tried to reproduce your successesful one:
1. added a step `conda install scipy`  before `pip install .` but turned out there are too many incompatibles and I can't `conda install scipy`
2. added a step `pip install scipy` before `pip install .` but I still got the `reduced` warning and `attribute error` (that make nonsense !!!) If I did `pythran scipy/stats/_count_paths_outside_method_pythran.py `, I will get no warning and error...
3. added a step `build another scipy from source and do conda develop .` before `pip install .` but I still got the `reduced` warning and `attribute error` .

That's really strange now, but I guess reproducing your successful one on my computer may be not important though.

**Comment by serge-sans-paille** [comment]
@rgommers I ended up considering it's not the end of the world if some optimization don't happen, and opt-in for a generic recovery system instead of what started to look like a patch over a patch. Can you confirm current version works fine?

**Comment by rgommers** [comment]
Indeed it does, see https://github.com/serge-sans-paille/pythran/pull/1810#issuecomment-868973390

**Comment by rgommers** [comment]
> That's really strange now, but I guess reproducing your successful one on my computer may be not important though.

It isn't I think, not 100% sure what's going on there. Maybe `pip install .` will fail where `python setup.py build_ext -i` works - because of build isolation. What happens when you do `pip install .` is the following:
- `pip` will see that SciPy has a `pyproject.toml` file.
- because of that, it will use "build isolation" - meaning it creates a _new_ virtualenv with only the build dependencies specified in `pyproject.toml`
- that then means you are back to the Pythran version from PyPI and no SciPy installed in that separate new virtualenv.

Python build & packaging is a little complicated!

**Comment by serge-sans-paille** [comment]
Fixed by #1810
